### ISSUE TITLE
Mantid nightly crash on running MARI reduction scripts.

### ISSUE THREAD

**Issue by mducle** [mantidproject/mantid#31501]
### Expected behavior

MARI reduction scripts run ok, and produces `nxspe` files and reduced workspaces.

### Actual behavior

Hard crash (C++ level `Unhandled Exception`) when the script is run.

### Steps to reproduce the behavior

On IDAaaS, run the attached scripts: 
[mari_reduction_files.zip](https://github.com/mantidproject/mantid/files/6495116/mari_reduction_files.zip)

If run on the commandline, it fails with: https://gist.github.com/mducle/6737d23436fe52757f9d339493f7badf

The key error seems to be:

```
********* UNHANDLED EXCEPTION *********
  what(): SpectrumInfo: No detectors for this workspace index. search object 0
```

### Platforms affected

IDAaaS (CentOS 7), others not checked.

**Comment by DanielMurphy22** [comment]
As mentioned on Slack, this issue may be caused by https://github.com/mantidproject/mantid/pull/30163

**Comment by DannyHindson** [comment]
I've found the cause of this - it is the PR mentioned above relating to changes to ConvertUnits and some of the Unit classes. The MARI reduction script performs a unit conversion from TOF to energy. Spectrum 0 in the workspace doesn't have any detectors and this used to result in the spectrum being masked but my changes had stopped this happening. I've restored the original behaviour and the script now runs fine for me. Will create a PR later on. The problem only affected the "Energy" unit.

I'm having a look at why non of the system tests failed because of this. I think it's because they all run with the "run_diagnostics" setting set to false whereas this is "True" in the MARI instrument parameter file.

**Comment by mducle** [comment]
Thanks @DannyHindson ! Yes, I think there is supposed to be a MARI reduction script/test in the system test and was wondering whether I've modified the one we used (which is hosted in the separate `scriptrepository` repo) too much compared to the systems test one. 

The MARI instrument definition is a bit weird in that spectrum 0, 4 and 28 don't actually physicall exist... it's been like this since forever (since before Mantid) so I didn't want to change it... I guess maybe they thought once upon a time to have additional detectors / monitors..

**Comment by DannyHindson** [comment]
@mducle - I've just created a PR for this issue. The existing system tests do go through the same code path as the script you supplied (including the run_diagnostics part - so my earlier comment suggesting this was a difference was wrong). The system test didn't catch the crash because the run numbers are different and the crash only happened inside MedianDetectorTest for certain data. So if the script where you found the problem is a fairly standard test script that you use then yes, worth us updating the Mantid system test to use the same run numbers. I haven't done this as part of the PR so suggest we do this as a follow up task

**Comment by martyngigg** [comment]
Fixed by #31517
